STORM-2204 Adding caching capabilities in HBaseLookupBolt

ambud opened the pull request:
https://issues.apache.org/jira/browse/STORM-2204

revans2 commented on external/storm-hbase/src/main/java/org/apache/storm/hbase/bolt/AbstractHBaseBolt.java:
```diff
@@ -37,8 +37,7 @@
 
-    protected OutputCollector collector;
-
+    protected transient OutputCollector collector;
```
Why is this transient?  I don't see us accessing it from multiple threads, and especially not changing it on multiple threads?

ambud replied:
Transient is for serializability errors, I am not sure why it didn't error out in earlier when people tried to use this code. 

Since Storm serializes the Bolt code for deployment this should be marked as transient.

revans2 replied:
You are right I got confused by the two.  Thanks for pointing that out :)

revans2 commented on external/storm-hbase/src/main/java/org/apache/storm/hbase/bolt/HBaseLookupBolt.java:
```diff
@@ -40,51 +48,85 @@
+	private HBaseValueMapper rowToTupleMapper;
+	private HBaseProjectionCriteria projectionCriteria;
+	private transient LoadingCache<byte[], Result> cache;
+	private transient boolean cacheEnabled;
```
Why are these transient? The bolt is not multi-threaded.

ambud replied:
Transient is for serializability, not to be confused with volatile.

revans2 commented on external/storm-hbase/src/main/java/org/apache/storm/hbase/bolt/HBaseLookupBolt.java:
```diff
@@ -40,51 +48,85 @@
  *
  */
 public class HBaseLookupBolt extends AbstractHBaseBolt {
```
Please don't use tab characters for indentation.  We are still working on the exact style guides, but we try to keep the style that is currently in the file and it was using spaces before and I expect the style guide to include spaces instead of tabs.

revans2 commented on external/storm-hbase/src/main/java/org/apache/storm/hbase/bolt/HBaseLookupBolt.java:
```diff
@@ -40,51 +48,85 @@
+				this.collector.reportError(e);
+				this.collector.fail(tuple);
+			}
+		}
```
There appears to be a lot of overlap in the code here, between the cache version and the non-cache version.  It would really be nice to at least combine most of this.  Alternatively we might want to use inheritance in some way to avoid the branching on the cacheEnabled config at all.

ambud commented on external/storm-hbase/pom.xml:
```diff
@@ -1,92 +1,92 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
That's happened because of my IDE. I would recommend adding the POM-Sort plugin to make sure this is auto fixed when the build runs. 

I took a crack at that a few months ago: https://github.com/srotya/srotya-parent/blob/master/pom.xml

ambud commented on external/storm-hbase/pom.xml:
```diff
@@ -1,92 +1,92 @@
+		<dependency>
+			<groupId>com.google.guava</groupId>
+			<artifactId>guava</artifactId>
+			<version>20.0</version>
```
Agree with the backwards, what are your thoughts on alternatives.

revans2 replied:
I have not really used too many alternatives, So I did a search

https://duckduckgo.com/?q=guava+cache+alternatives&t=ffab&ia=qa

Both https://cache2k.org/ and http://www.ehcache.org/ look interesting, but I have not actually used any of them

revans2 commented:
Overall it looks fine to me now.  I am still a bit concerned about guava as a dependency, but beyond that it seems like a great addition to the bolt.

ben-manes commented:
@revans2 perhaps [Caffeine](https://github.com/ben-manes/caffeine)?

HeartSaVioR commented:
I agree it's still better to not relying on Guava since storm-core is shading Guava but external modules are not. 
Btw, I didn't use other cache modules too, but ehcache seems to be well-known and widely-used cache library, and Caffeine seems to be a drop-in replacement for Guava cache.

@ben-manes I guess you're the author of Caffeine. Could you introduce Caffeine?

HeartSaVioR commented:
Forgot to comment for other side. It looks great.

ben-manes commented:
I was a co-author of Guava's cache, too. 

Guava had originally considered soft references an ideal caching scheme, since they offer great concurrency and GC is for memory management. That evolved from `ReferenceMap` to `MapMaker` to optimize space, especially in regards to computations (no need for a `Future` wrapper). Unfortunately soft references result in awful performance outside of a micro-benchmark due to causing full GCs. In parallel, I had been experimenting with approaches for a concurrent LRU cache ([CLHM](https://github.com/ben-manes/concurrentlinkedhashmap)) and when joining Google helped to retrofitted its ideas onto Guava. There was a lot of good that came out of that, but I left before working on optimizing for performance.

Java 8 provided an excuse to start from scratch. Caffeine is much faster and packs in even more features. I also spent time exploring eviction policies, which led to co-authoring a paper on a new technique called TinyLFU. That has a near optimal hit rate, low memory footprint, and amortized O(1) overhead. This is done by tracking frequency in a popularity sketch. The same concurrency model in CLHM and Guava is used (inspired by a write-ahead log), which allows for concurrent O(1) reads and writes.

The [HighScalability article](http://highscalability.com/blog/2016/1/25/design-of-a-modern-cache.html) provides an overview of the algorithms that I use.

vesense commented:
Overall looks good to me. My major concern is that on-heap caches(like Guava cache, Ehcache)  might cause bad GC situations.  I didn't use Caffeine in the past, maybe is a candidate.

revans2 commented:
Caffeine sounds like a great alternative to Guava and also seems to address some GC concerns.  

@vesense I agree storing any large amount of data on heap will impact GC, we do that all the time with all of the queues that we have.  I think for the most part as long as GC is tuned properly for the topology, we don't get a lot of full GCs causing promotion in the cache, and TTL in the cache is not too long it will be fine.  If we do run into serious GC issues we can then look at off heap cacheing.

vesense commented:
@revans2 Yes, we can find a balance between high rate of cache hits and full GCs. I'm OK for adding a built-in cache if we set parameters carefully.

revans2 commented:
@vesense it is off by default so it would be enabled/tuned on a per topology basis.

ambud commented:
@revans2 @vesense @ben-manes Thanks for the valuable feedback, I am looking into integration with Caffeine. Should have a new commit soon.

@revans2 Indeed the intention is to have the cache tuned based on topology. The ideal use case is where caching actually makes sense and certain keys may have a higher hit frequency then others.

Ideally this bolt should be used with FieldGrouping when caching is enabled so cache misses minimal. Something to be kept in mind is distribution of keys (executor balancing).

vesense commented:
+1

vesense commented on external/storm-hbase/src/main/java/org/apache/storm/hbase/bolt/HBaseLookupBolt.java:
```diff
@@ -40,51 +48,81 @@
+                  @Override
+                  public Result load(byte[] rowKey) throws Exception {
+                     Get get = hBaseClient.constructGetRequests(rowKey, projectionCriteria);
+                     LOG.debug("Cache miss for key:"+new String(rowKey));
```
@ambud
It would be better to use `Cache miss for key: {}` instead of `+`. After fixing this, you can squash commits into one.

HeartSaVioR commented:
One thing to consider is that Caffeine seems to require Java 8, which means that this patch can't be shipped to 1.x version line. Do we want to keep using Guava for 1.x branch?
@ben-manes Is there any versions for Caffeine which works with Java 7?

ben-manes commented:
Nope. Sorry, since your compilation target is 1.8 I hadn't thought you'd need that.

ambud commented:
So should I revert back my changes; seems like the build is currently failing; additional tests pushed by upstream?

HeartSaVioR commented:
@revans2 What do you think about this? I'm in favor of adopting Caffeine, and I'm even OK to use Caffeine to master and Guava to 1.x.

revans2 commented:
@HeartSaVioR I also am find with Caffeine and even Caffeine on 1.x (assuming it is off by default)

revans2 commented:
+1 the code looks fine to me.

HeartSaVioR commented:
@ambud 
The code looks good except what @vesense commented. 

Two things more to address:

1. It would be better to document new configurations. Without documentation, end-users have no idea about added feature. `external/storm-hbase/README.md` and `docs/storm-hbase.md`.

2. The code already uses JDK 8 API (Map.getOrDefault()), so can't get it as it is for 1.x. Could you provide a new PR for 1.x branch?
It would be also great if you can test it (with Caffeine) on JRE7 (expected to not work but we can document the precondition for JRE version) and JRE8 (expected to work).
cc. @ben-manes Is my expectation right?

Thanks in advance!

ben-manes commented:
Should fail, e.g. Class version error.

ptgoetz commented:
+1, but I agree with @HeartSaVioR that the new cache configuration options need to be documented before this is merged.

ambud commented:
Added documentation to readme file.

Fixed the debug logging by using isDebugEnabled checks

Are we going to use Guava caching implementation for 1.x? @HeartSaVioR @revans2

HeartSaVioR commented:
@ambud I'm OK to use Guava for 1.x. IMHO it would be better to provide complete set of features for guaranteed environment (JDK version) instead of leaving 'warn' to documentation. 
And some other external modules also use Guava, too.
@revans2 Are you OK to use Guava only for 1.x? Or you would like to keep using Caffeine for also 1.x and document it?

ambud commented:
Caffeine is JDK 8 only so won't work for 1.x, since JDK 7 compilation will be tested.

HeartSaVioR commented:
@ambud 
Yeah if we add unit test for the feature, unit test will fail on JDK 7 which will be always failing on Travis CI. Nice catch. Could you address this to use Guava for 1.x?

ambud commented:
Sure thing, I will add the Guava patch code originally added for this to 1.x and 0.10.x branches

vesense commented:
+1 for merging to master after updating `docs/storm-hbase.md`. `external/storm-hbase/README.md` has been documented.

revans2 commented:
I guess guava is OK for 1.x.  I would prefer to see it shaded if we do go with Guava, but I am only a -0 if it is not shaded.

HeartSaVioR commented:
OK. Since we have other modules which depends on Guava, it might be better to have a rule and apply all of them. +1 to shade Guava on external modules if possible.

ambud commented:
Can we merge this? @HeartSaVioR @revans2 @vesense 

I will need to open another pull request for 1.x branch.

ambud commented:
1.x branch PR https://github.com/apache/storm/pull/1810

ambud commented:
Can we merge this @revans2 @HeartSaVioR @vesense ?

vesense commented:
@ambud Sorry for the delay response. Before we merge this in, can you replace all the tab space to white space?

ambud commented:
Done

vesense commented:
Thanks @ambud Squashed and merged into master. And I added you as the contributor.

ambud commented:
Thank you @vesense. Could we merge the 1.x pull request as well?

vesense commented:
@ambud I'll wait for other committers to vote for it before merging into 1.x-branch. And this may take some time.
